pkg/config/polling_manager_test.go

```diff
@@ -44,68 +46,184 @@ func newExecGroup() *utils.ExecGroup {
+
+	// poll after 100ms
+	eg.Go(configManager.Start)
+	evaluationMethod := func() bool {
+		_, err := configManager.GetConfig()
+		return err != nil
+	}
+	assertPeriodically(t, evaluationMethod)
+	mockRequester.AssertExpectations(t)
+	eg.TerminateAndWait()
+}
+
 func TestNewPollingProjectConfigManagerWithSimilarDatafileRevisions(t *testing.T) {
 	mockDatafile1 := []byte(`{"revision":"42","botFiltering":true}`)
 	mockDatafile2 := []byte(`{"revision":"42","botFiltering":false}`)
 	projectConfig1, _ := datafileprojectconfig.NewDatafileProjectConfig(mockDatafile1)
 	mockRequester := new(MockRequester)
-	mockRequester.On("Get", []utils.Header(nil)).Return(mockDatafile1, http.Header{}, http.StatusOK, nil)
+	mockRequester.On("Get", []utils.Header(nil)).Return(mockDatafile2, http.Header{}, http.StatusOK, nil)
 
 	sdkKey := "test_sdk_key"
-
 	eg := newExecGroup()
-	configManager := NewPollingProjectConfigManager(sdkKey, WithRequester(mockRequester))
-	eg.Go(configManager.Start)
-	mockRequester.AssertExpectations(t)
+	configManager := NewPollingProjectConfigManager(sdkKey, WithRequester(mockRequester), WithInitialDatafile(mockDatafile1), WithPollingInterval(100*time.Millisecond))
 
+	var numberOfCalls uint64 = 0
+	callback := func(notification notification.ProjectConfigUpdateNotification) {
+		atomic.AddUint64(&numberOfCalls, 1)
+	}
+	id, _ := configManager.OnProjectConfigUpdate(callback)
+	assert.NotEqual(t, 0, id)
+
+	// initialized with hardcoded datafile
 	actual, err := configManager.GetConfig()
 	assert.Nil(t, err)
 	assert.NotNil(t, actual)
 	assert.Equal(t, projectConfig1, actual)
 
-	configManager.SyncConfig(mockDatafile2)
-	actual, err = configManager.GetConfig()
+	// poll after 100ms
+	eg.Go(configManager.Start)
+	evaluationMethod := func() bool {
+		actual, _ = configManager.GetConfig()
+		return reflect.DeepEqual(projectConfig1, actual)
+	}
+	assertPeriodically(t, evaluationMethod)
+	mockRequester.AssertExpectations(t)
+	// Check no notification was sent for similar datafile revision
+	assert.Equal(t, uint64(0), atomic.LoadUint64(&numberOfCalls))
+	eg.TerminateAndWait()
+}
+
+func TestNewAsyncPollingProjectConfigManagerWithSimilarDatafileRevisions(t *testing.T) {
+	mockDatafile1 := []byte(`{"revision":"42","botFiltering":true}`)
+	mockDatafile2 := []byte(`{"revision":"42","botFiltering":false}`)
+	projectConfig1, _ := datafileprojectconfig.NewDatafileProjectConfig(mockDatafile1)
+	mockRequester := new(MockRequester)
+	mockRequester.On("Get", []utils.Header(nil)).Return(mockDatafile2, http.Header{}, http.StatusOK, nil)
+
+	sdkKey := "test_sdk_key"
+	eg := newExecGroup()
+	asyncConfigManager := NewAsyncPollingProjectConfigManager(sdkKey, WithRequester(mockRequester), WithInitialDatafile(mockDatafile1), WithPollingInterval(100*time.Millisecond))
+
+	var numberOfCalls uint64 = 0
+	callback := func(notification notification.ProjectConfigUpdateNotification) {
```

Comment:
if we have similar datafile revision, then how it's gonna be called. please explain the reason to declare here.